src/builder/QuarkBuilder.sol

```diff
@@ -70,50 +122,93 @@ contract QuarkBuilder {
-    // by making it equivalent in structure to data already in portfolios.
-    struct AssetPositions {
+    // Note: This is just the AssetPositions type with an extra `chainId` field
+    struct AssetPositionsWithChainId {
+        uint256 chainId;
         address asset;
         string symbol;
         uint256 decimals;
         uint256 usdPrice;
         AccountBalance[] accountBalances;
     }
 
-    struct AccountBalance {
-        address account;
-        uint256 balance;
+    // TODO: convert strings to lower case before comparing. maybe rename to `compareSymbols`
+    function compareStrings(string memory a, string memory b) internal pure returns (bool) {
+        return keccak256(abi.encodePacked((a))) == keccak256(abi.encodePacked((b)));
     }
 
-    struct Payment {
-        bool isToken;
-        string currency;
-        uint256[] chainIds;
-        uint256[] maxCosts;
+    function filterAssetPositionsForSymbol(string memory assetSymbol, ChainAccount[] memory chainAccountsList) internal pure returns (AssetPositionsWithChainId[] memory) {
+        uint numMatches = 0;
+        // First loop to count the number of matching AssetPositions
+        for (uint i = 0; i < chainAccountsList.length; ++i) {
+            for (uint j = 0; j < chainAccountsList[i].assetPositionsList.length; ++j) {
+                if (compareStrings(chainAccountsList[i].assetPositionsList[j].symbol, assetSymbol)) {
+                    numMatches++;
+                }
+            }
+        }
+
+        AssetPositionsWithChainId[] memory matchingAssetPositions = new AssetPositionsWithChainId[](numMatches);
+        uint index = 0;
+        // Second loop to populate the matchingAssetPositions array
+        for (uint i = 0; i < chainAccountsList.length; ++i) {
+            for (uint j = 0; j < chainAccountsList[i].assetPositionsList.length; ++j) {
+                if (compareStrings(chainAccountsList[i].assetPositionsList[j].symbol, assetSymbol)) {
+                    AssetPositions memory assetPositions = chainAccountsList[i].assetPositionsList[j];
+                    matchingAssetPositions[index] = AssetPositionsWithChainId({
+                        chainId: chainAccountsList[i].chainId,
+                        asset: assetPositions.asset,
+                        symbol: assetPositions.symbol,
+                        decimals: assetPositions.decimals,
+                        usdPrice: assetPositions.usdPrice,
+                        accountBalances: assetPositions.accountBalances
+                    });
+                    index++;
+                }
+            }
+        }
+
+        return matchingAssetPositions;
     }
 
-    function getPriceFeed(string assetSymbol, uint256 chaindId) internal pure (address) {
+    function getAssetPositionForSymbolAndChain(string memory assetSymbol, uint256 chainId, ChainAccount[] memory chainAccountsList) internal pure returns (AssetPositionsWithChainId memory) {
```

Comment:
Nit: there can be multiple positions for the same `{symbol, chain id}`. Multiple wallets.

Reply:
Good catch, left out the `s` in `Positions`. Done